### PR TITLE
fix "IDO3" Use of Assert instead of throw

### PR DESCRIPTION
**Pull request by Lichen9618**

### CODE REVIEW

File: contracts/IdoContract/src/IdoContract.cs [hunk 1]
```diff
@@ -150,10 +148,10 @@ public static bool Unstake(UInt160 userAddress, BigInteger unstakeAmount)
         {
             UInt160 assetHash = GetStakeAssetHash();
             BigInteger amountBefore = GetBalanceOfToken(assetHash, Runtime.ExecutingScriptHash);
-            if (!Runtime.CheckWitness(userAddress)) throw new Exception("CUWF"); //check user witness fail
+            ExecutionEngine.Assert(Runtime.CheckWitness(userAddress), "check user witness fail");
             UserStakeInfo stakeInfo = GetUserStakeInfo(userAddress);
-            if (stakeInfo.isNewUser) throw new Exception("bad address");
-            if (stakeInfo.lastStakeAmount < unstakeAmount) throw new Exception("bad amount");
+            ExecutionEngine.Assert(!stakeInfo.isNewUser, "bad userAddress");
+            ExecutionEngine.Assert(stakeInfo.lastStakeAmount > unstakeAmount, "bad amount");
```
Comment (shargon) [inline contracts/IdoContract/src/IdoContract.cs:154]: `>=`

File: contracts/IdoContract/src/IdoContract.cs [hunk 2]
```diff
@@ -456,10 +454,11 @@ private static BigInteger GetUserClaimAmountImple(byte[] key)
         public static bool ClaimToken(UInt160 user, UInt160 idoPairContractHash)
         {
             RegisteredProject project = GetRegisteredProject(idoPairContractHash);
-            if (!project.isReviewed || Ledger.CurrentIndex - project.reviewedHeight < 2 * GetVoteTimeSpan()) throw new Exception("RNE"); // project reviewed not end yet
+            ExecutionEngine.Assert(project.isReviewed && Ledger.CurrentIndex - project.reviewedHeight >= 2 * GetVoteTimeSpan(), "project review not end");
             byte[] key = GetUserClaimAmountKey(idoPairContractHash, user);
             BigInteger oldAmount = GetUserClaimAmountImple(key);
-            if (oldAmount <= 0) throw new Exception("NCT"); // no unclaimed token
+            ExecutionEngine.Assert(oldAmount > 0, "no unclaimed token");
+            SafeTransfer(project.tokenHash, Runtime.ExecutingScriptHash, user, oldAmount);
```
Comment (shargon) [inline contracts/IdoContract/src/IdoContract.cs:461]: Double transfer? it's already done in L464

File: contracts/IdoContract/src/IdoContract.cs [hunk 3]
```diff
@@ -468,19 +467,19 @@ public static bool ClaimToken(UInt160 user, UInt160 idoPairContractHash)
...
+            ExecutionEngine.Assert(amount > 0, "bad swap amount");
             //transfer asset part
             BigInteger balanceBefore = GetBalanceOfToken(project.tokenHash, idoPairContractHash);
             BigInteger spendAssetAmount = project.tokenOfferingPrice * amount / PriceDenominator;
             SwapAsset(user, amount);
             CallSwap(idoPairContractHash);
             SafeTransfer(GetSpendAssetHash(), user, idoPairContractHash, spendAssetAmount);
-            BigInteger balanceAfter = GetBalanceOfToken(project.tokenHash, idoPairContractHash);
-            if (balanceBefore - balanceAfter != amount) throw new Exception("AMC"); // amount not correct
+            BigInteger balanceAfter = GetBalanceOfToken(project.tokenHash, user);
```
Comment (shargon) [inline contracts/IdoContract/src/IdoContract.cs:481]: Why `user` ?

Comment (Lichen9618) [reply contracts/IdoContract/src/IdoContract.cs:481]: should be idoPairContractHash

File: contracts/IdoContract/src/IdoContract.cs (hunk 1 above)
Comment (Lichen9618) [reply contracts/IdoContract/src/IdoContract.cs:154]: right

File: contracts/IdoContract/src/IdoContract.cs (hunk 2 above)
Comment (Lichen9618) [reply contracts/IdoContract/src/IdoContract.cs:461]: keep L464 and remove L461

**Pull request merged**